benchmarks/threadsafe_let_block.rb

```diff
@@ -0,0 +1,304 @@
+    MemoizedHelpers.module_eval do
+      def subject
+        __memoized.fetch(:subject) do
+          __memoized[:subject] = begin
+            described = described_class || self.class.metadata.fetch(:description_args).first
+            Class === described ? described.new : described
+          end
+        end
+      end
+
+      def __memoized
+        @__memoized ||= {}
+      end
+    end
+
+    class << MemoizedHelpers::ContextHookMemoized
+      alias fetch for
+    end
+
+    MemoizedHelpers::ClassMethods.module_eval do
+      def let(name, &block)
+        # We have to pass the block directly to `define_method` to
+        # allow it to use method constructs like `super` and `return`.
+        raise "#let or #subject called without a block" if block.nil?
+        MemoizedHelpers.module_for(self).__send__(:define_method, name, &block)
+
+        # Apply the memoization. The method has been defined in an ancestor
+        # module so we can use `super` here to get the value.
+        if block.arity == 1
+          define_method(name) { __memoized.fetch(name) { |k| __memoized[k] = super(RSpec.current_example, &nil) } }
+        else
+          define_method(name) { __memoized.fetch(name) { |k| __memoized[k] = super(&nil) } }
+        end
+      end
+    end
+  end
+
+
+  def use_threadsafe
+    return if @state == :new
+    @state = :new
+    MemoizedHelpers.module_eval do
+      def subject
+        __memoized.for(:subject) do
+          described = described_class || self.class.metadata.fetch(:description_args).first
+          Class === described ? described.new : described
+        end
+      end
+
+      def __memoized
+        @__memoized ||= Memoized.new
+      end
+    end
+
+    MemoizedHelpers::ClassMethods.module_eval do
+      def let(name, &block)
+        # We have to pass the block directly to `define_method` to
+        # allow it to use method constructs like `super` and `return`.
+        raise "#let or #subject called without a block" if block.nil?
+        MemoizedHelpers.module_for(self).__send__(:define_method, name, &block)
+
+        # Apply the memoization. The method has been defined in an ancestor
+        # module so we can use `super` here to get the value.
+        if block.arity == 1
+          define_method(name) { __memoized.for(name) { super(RSpec.current_example, &nil) } }
+        else
+          define_method(name) { __memoized.for(name) { super(&nil) } }
+        end
+      end
+    end
+  end
+end
+
+
+require 'benchmark/ips'
+
+
+# Given
+#   let(:name) { nil }
+# how many iterations per second for
+#   my branch
+#   master
+# when called
+#   once
+#   10x
+# Same thing with a call to super
+
+
+helper = LetBlockImplementationHelper.new
+
+
+puts helper.title "versions"
+puts "RUBY_VERSION             #{RUBY_VERSION}"
+puts "ruby -v                  #{`ruby -v`}"
+puts "Benchmark::IPS::VERSION  #{Benchmark::IPS::VERSION}"
+puts "rspec-core SHA           #{`git log --pretty=format:%H -1`}"
```

Comment:
I had never thought about including the SHA in the benchmark output...cool idea!  I'll probably steal this in the future for other benchmarks :).